### DISCUSSION TITLE
Document cookie header preservation

### DISCUSSION THREAD

**Discussion by aaronadamsCA** [remix-run/remix#5998] [Proposals]
I learned something new and important today: If you add `Set-Cookie` headers in a parent route, they are automatically preserved in child routes!

This was an intentional feature added more than two years ago: https://github.com/remix-run/remix/commit/d37c00508b671758c714b4c7672280d8e8de08ea

However it seems it was never documented. It really should be, as it's incredibly useful.

After I read the documentation, which told me I should set my headers in my leaf route responses only, I dutifully added my session cookie header to every single leaf route. I also accidentally added it to a layout route; imagine my surprise to find I was getting two duplicate session cookie headers!

Setting cookies in my root route only is very convenient and definitely something I will be using. This should be documented and specifically called out, as it's very useful!

**Comment by sergiodxa** [comment]
There's a reason to still want to only set cookies in leaf routes, if both parent and child/leaf routes sets the same cookie (e.g. the session cookie) only the last one to resolve will be preserved.

It's possible, [with some changes on your HTTP server](https://sergiodxa.com/tutorials/automatically-commit-session-on-remix) to solve it most of the time, but it can always happen on a client-side navigation.

**Reply by aaronadamsCA** [reply]
I think that's probably important as a "gotcha" specific to `cookieSessionStorage`. If you're using any other session storage, it probably makes a lot more sense to set the cookie header at the root route only, since the key won't change after it's created.
